### PR TITLE
View notification threads in Octobox

### LINKED ISSUE #1108: Show conversation threads for issues
Quick todo list for myself:

- [x] Add XXL and XXXL mixin for displays
- [x] Render with sidebar shown on /index
- [x] Render with sidebar hidden in /show
- [x] Toggle menu on click of subject to thread view and hide menu
- [x] Back button should not be back
- [x] Add btn styles for merged et al
- [x] Add PR statuses on threads themselves
- [ ] Update currentlySelectedRow in line with changes in selected notification
- [x] Remove message preview is archived/deleted (and update rows)
- [x] Make sure there's something rendered for notifications with no discussions
- [x] Fix viewing threads in archive et al
- [x] add user option for viewing threads in Octobox or GH
- [x] create responsive (mobile) layouts for threads viewed in Octobox
- [x] amend layout (moving back down a breakpoint) if user is viewing threads on GitHub

### PR DESCRIPTION
**Pull request by BenJam**
Adds threaded views to Octobox for any `subjectable?` notification. 
Adds the above as a user-opt in feature available through `/settings`

Will be removed if people don't use it... 

Closes #1108

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -56,8 +60,17 @@ $(document).on('click', 'a.js-sync', function(e) {
...
-})
+});
+
+$(document).on('click', '.subject-link', function () {
+  history.pushState({thread: $(this).attr('href')}, 'Octobox', $(this).attr('href'))
+  $.get($(this).attr('href'), function(data){
+    $('#thread').html(data)
+  });
+  Octobox.openThread();
+  return false;
```
Comment (andrew) [inline app/assets/javascripts/application.js]: We should probably move this to octobox.js

File: app/assets/javascripts/octobox.js [hunk 2]
```diff
@@ -479,19 +539,25 @@ var Octobox = (function() {
...
+    muteSelected: muteSelected,
     markReadSelected: markReadSelected,
-    archive: archive,
-    unarchive: unarchive,
+    archiveSelected: archiveSelected,
+    unarchiveSelected: unarchiveSelected,
     toggleSelectAll: toggleSelectAll,
     sync: sync,
     markRowCurrent: markRowCurrent,
+    closeThread: closeThread,
```
Comment (andrew) [inline app/assets/javascripts/octobox.js]: Would be good to have a keyboard shortcut for closing the thread, perhaps `esc`?

File: app/controllers/notifications_controller.rb [hunk 3]
```diff
@@ -90,12 +90,49 @@ def index
...
+      @assigned              = scope.reorder(nil).assigned(current_user.github_login).count
+      @visiblity             = scope.reorder(nil).distinct.joins(:repository).group('repositories.private').count
+      @repositories          = Repository.where(full_name: scope.reorder(nil).distinct.pluck(:repository_full_name)).select('full_name,private')
+    end
+
+    scope = current_notifications(scope)
+    check_out_of_bounds(scope)
+
+    @unread_count = user_unread_count
+    @display_comments = current_user.display_comments
```
Comment (andrew) [inline app/controllers/notifications_controller.rb]: This line can be removed, instead using `display_thread?` in the view
```suggestion

```

File: app/controllers/notifications_controller.rb [hunk 4]
```diff
@@ -90,12 +90,49 @@ def index
     check_out_of_bounds(scope)
 
     @unread_count = user_unread_count
+    @display_comments = current_user.display_comments
```
Comment (andrew) [inline app/controllers/notifications_controller.rb]:
```suggestion

```

File: app/helpers/notifications_helper.rb [hunk 5]
```diff
@@ -273,4 +289,36 @@ def sidebar_notification_status(status)
...
+  end
+
+  def notification_link(notification)
+    display_thread? && notification.subjectable? ? notification_url(notification, filtered_params) : notification.web_url
+  end
+
+  def display_thread?
+    Octobox.include_comments? && current_user.display_comments
+  end
+
```
Comment (andrew) [inline app/helpers/notifications_helper.rb]:
```suggestion

```

File: app/views/notifications/index.html.erb [hunk 6]
```diff
@@ -8,109 +8,20 @@
...
-            <% end %>
-            <% if current_user.last_synced_at %>
-              <p class='text-muted'>
-                <small>Last sync: <%= time_ago_in_words current_user.last_synced_at %> ago</small>
-              </p>
-            <% end %>
-          </div>
+    <%= render "list" %>
+  </div>
+  <% if @display_comments %>
```
Comment (andrew) [inline app/views/notifications/index.html.erb]:
```suggestion
  <% if display_thread? %>
```

File: app/assets/javascripts/octobox.js (hunk 2 above)
Comment (BenJam) [reply app/assets/javascripts/octobox.js]: 👍

File: app/assets/javascripts/application.js (hunk 1 above)
Comment (BenJam) [reply app/assets/javascripts/application.js]: moving this (wrapping a pushLink function into octobox.js with this logic inside and passing $this to it) resulted in a strange xhr deprecation warning and a refusal to open the link...

Comment (andrew) [reply app/assets/javascripts/application.js]: Fixed in https://github.com/octobox/octobox/pull/1169/commits/5c26b637afe44d3401e79f9fd8571ed8e64da5d2

**Comment by andrew** [comment]
There are over 1000 subjects in Octobox with more than 20 comments, one of them with 380 comments, so we should definitely add some form of pagination soon.

**Comment by andrew** [comment]
screen shot 2018-10-31 at 13 22 26
 
Comments from bot users have broken avatars

**Comment by andrew** [comment]
When there are errors whilst rendering the thread partial, the UI doesn't show that, probably need a proper error handler on the ajax call.

**Comment by andrew** [comment]
There's a few errors when trying to show a thread for a notification that doesn't have a subject, we should probably only link to the thread view if it has a subject.

**Comment by andrew** [comment]
We might want to add a way to visually show what's going to happen when clicking on a notification in the list, for example:

screen shot 2018-10-31 at 13 37 05

One of these commits has a subject synced and will open the thread view, and the other will open a new tab with the commit on GitHub, but you have no idea which will happen until you click.

**Pull request merged**